### DISCUSSION TITLE
How Does `loaderHeaders` Work?

### DISCUSSION THREAD

**Discussion by fweth** [remix-run/remix#2005] [Q&A] [answered]
[EDIT] I think I understand now, but I can't delete the question(?)  Anyway, this gist was helpful: https://gist.github.com/ryanflorence/ddc5604a8ae9e068ef4e4478e8fa845a

Beginner question here, feel free to delete.  I'm new to HTTP headers and all that stuff, I don't understand how `headers` is able to access `loaderHeaders` when the `loader` runs async on the server?  So my understanding is, you request some URL, Remix serves the URL with a set of headers from `headers` and when the `loader` finishes, the data gets sent in an extra package with possible different headers.  But how can the headers from `loader` influence the headers from `headers`, which need to be fixed at response time?

**Comment by ccssmnn** [comment, marked as answer]
`headers` describe the headers which are sent when your route is called.

The headers you send in your `loader`'s response are sent when your loader is called directly. E.g. when using Remix `<Link/>` component for navigation.

You might want to change the headers based on the information in the loader, which is why you can use the `loaderHeaders` within your `headers` function.

**Reply by fweth** [reply]
OK, but the `loderHeaders` only exist when I navigate on the page, so I should add a fallback in case the user directly requests the given route?

[EDIT] Or can the headers be changed after the initial response?  Like telling the browser, hey, maybe you shouldn't store this site in the cache as long as we told you initially?

**Reply by ccssmnn** [reply]
You should use both. JavaScript could fail and `<Link/>` falls back to normal links. Also users could navigate to the route directly.

**Reply by ccssmnn** [reply]
You cannot change the response headers after the response is sent.

**Reply by fweth** [reply]
Thanks a lot for the prompt answer!
